lab/skel_graph/shortest_path/dijkstra/main.cpp

```diff
@@ -1,157 +1,179 @@
+
+// Dijkstra
+// O(m log n)
+// Deoarece folosesc heapuri binare (priority_queue),
+// complexitatea este cea de mai sus
+
+// http://www.infoarena.ro/problema/dijkstra
+
+#include <bits/stdc++.h>
+
+#define NMAX 50010   // ATENTIE la cat e in problema curenta !!!
 #define oo (1 << 30) // ATENTIE la cat e in problema curenta !!!
+#define NO_PARENT (-1)
+
 using namespace std;
 
-// tipul edge
-// muchia de cost "cost", de la node la neighbour,
-// va fi un element edge(neighbour, cost) in G[node]
-typedef pair<int, int> edge;
-#define neighbour first
-#define cost second
-
-// aceleasi semnificatii ca la BFS
-int n, m;
-int d[NMAX], p[NMAX];
-vector<edge> G[NMAX];
-
-// declararea unui priority queue in care pun numere intregi (noduri)
-// si cu un criteriu custom de prioritate
-
-// varianta 1
-// http://www.cplusplus.com/reference/queue/priority_queue/priority_queue/
-struct cmp {
-  // returneaza true daca x ar trebui "sa fie sub" y in arbore
-  bool operator()(const int &x, const int &y) const { return d[x] > d[y]; }
-};
-priority_queue<int, vector<int>, cmp> pq;
+class Task {
+public:
+  void solve() {
+    read_input();
+    get_result();
+    print_output();
+  }
 
-// varianta 2 - C++11 :D
-// http://en.cppreference.com/w/cpp/language/decltype
-// bool cmp(const int &x, const int &y) {
-//     // returneaza true daca x ar trebui "sa fie sub" y in arbore
-//     return d[x] > d[y];
-// }
-// priority_queue<int, vector<int>, decltype(&cmp)> pq(cmp);
+private:
+  // n = numar de noduri, m = numar de muchii
+  int n, m;
 
-void read_input() {
-  cin >> n >> m;
+  int source;
 
-  for (int i = 1; i <= m; ++i) {
-    int x, y, c; // x -> y, de cost c
-    cin >> x >> y >> c;
+  // adj[i] contine perechi de tipul : vecin , costul muchiei i -> vecin
+  vector<pair<int, int>> adj[NMAX];
 
-    G[x].push_back(edge(y, c));
-    // G[y].push_back(edge(x, c)); // sterg daca e orientat !!!
-  }
-}
+  // say something
+  priority_queue<pair<int, int>, vector<pair<int, int>>,
+                 std::greater<pair<int, int>>>
+      pq;
+
+  // d[i] = distanta minima de la de source la nodul i
+  vector<int> d;
+
+  // p[i] = parintele lui i de pe drumul minim de la source la nodul i
+  vector<int> p;
 
-void Dijkstra(int source) {
-  // ETAPA 1
-  for (int i = 1; i <= n; ++i) {
-    d[i] = oo;
-    p[i] = oo;
+  void read_input() {
+    cin >> n >> m;
+
+    d.resize(n + 1);
+
+    p.resize(n + 1);
+
+    // pe infoarena sursa este 1
+    // modificati linia urmatoare pentru a seta alta sursa
+    source = 1;
+
+    for (int i = 1; i <= m; ++i) {
+      // citim muchii x -> y, de cost c
+      int x, y, c;
+      cin >> x >> y >> c;
+
+      // adaugam in lista de adiacenta a nodului x
+      // nodul y cu costul muchiei x -> y = c
+      adj[x].push_back({y, c});
+    }
   }
 
-  // ETAPA 2
-  pq.push(source);
-  d[source] = 0;
-  p[source] = 0;
+  void get_result() { Dijkstra(source); }
+
+  void Dijkstra(int source) {
```

Comment:
please change method to
```
void Dijkstra(int source, vector<int> &d, vector<int> &p);
```

It's a proper way to reuse this function (all inputs, all outputs - exluding the graf).